File: backend/src/schemas/grade.py
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime


class GradeCreate(BaseModel):
    student_id: UUID
    exam_name: str = Field(..., min_length=1, max_length=255)
    subject: str = Field(..., min_length=2, max_length=100)
    score: float = Field(..., ge=0)
    max_score: float = Field(default=100.0, gt=0)
    wrong_questions: list[int] | None = None
    notes: str | None = None


class GradeCSVRow(BaseModel):
    student_code: str
    exam_name: str
    subject: str
    score: float
    max_score: float = 100.0


class GradeBulkUpload(BaseModel):
    subject: str = Field(..., min_length=2, max_length=100)
    exam_name: str = Field(..., min_length=1, max_length=255)
    grades: list[GradeCSVRow]


class GradeResponse(BaseModel):
    id: UUID
    student_id: UUID
    teacher_id: UUID
    exam_name: str
    subject: str
    score: float
    max_score: float
    wrong_questions: list[int] | None = None
    notes: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AIAnalysisRequest(BaseModel):
    student_id: UUID
    subject: str = Field(..., min_length=2, max_length=100)
    grades: list[dict] = Field(
        ...,
        description='List of grade objects like [{"exam": "quiz1", "score": 65, "wrong_questions": [3,7,12]}]',
    )


class AIAnalysisResponse(BaseModel):
    strengths: list[str]
    weaknesses: list[str]
    recommended_focus: list[str]
    next_exercise_suggestion: str

File: teacher_mobile/lib/core/models/grade.dart
class GradeModel {
  final String id;
  final String studentId;
  final String? studentCode;
  final String examName;
  final String subject;
  final double score;
  final double maxScore;
  final List<int>? wrongQuestions;
  final String? notes;
  final DateTime createdAt;

  const GradeModel({
    required this.id,
    required this.studentId,
    this.studentCode,
    required this.examName,
    required this.subject,
    required this.score,
    required this.maxScore,
    this.wrongQuestions,
    this.notes,
    required this.createdAt,
  });

  factory GradeModel.fromJson(Map<String, dynamic> json) {
    return GradeModel(
      id: json['id']?.toString() ?? '',
      studentId: json['student_id']?.toString() ?? '',
      studentCode: json['student_code']?.toString(),
      examName: json['exam_name']?.toString() ?? '',
      subject: json['subject']?.toString() ?? '',
      score: (json['score'] ?? 0).toDouble(),
      maxScore: (json['max_score'] ?? 100).toDouble(),
      wrongQuestions: json['wrong_questions'] != null
          ? List<int>.from(json['wrong_questions'] as List)
          : null,
      notes: json['notes']?.toString(),
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'student_id': studentId,
        'student_code': studentCode,
        'exam_name': examName,
        'subject': subject,
        'score': score,
        'max_score': maxScore,
        'wrong_questions': wrongQuestions,
        'notes': notes,
        'created_at': createdAt.toIso8601String(),
      };

  double get percentage => maxScore > 0 ? (score / maxScore) * 100 : 0;
}
